ami/models/latent_visualization_decoder.py

```diff
@@ -0,0 +1,252 @@
+                Input features.
+                (shape: [batch_size, in_channels, height, width])
+        Returns:
+            torch.Tensor:
+                Output features.
+                (shape: [batch_size, out_channels, height, width])
+        """
+        h = self.in_layers(x)
+        h = self.out_group_norm(h)
+        h = self.out_layers(h)
+        return self.skip_connection(x) + h
+
+
+class DecoderBlock(nn.Module):
+    """A block for decoder."""
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        n_res_blocks: int,
+        use_upsample: bool,
+    ) -> None:
+        """A block for LatentVisualizationDecoderr.
+
+        Args:
+            in_channels (int):
+                Input num of channels.
+            out_channels (int):
+                Output num of channels.
+            n_res_blocks (int):
+                Num of resblocks in this block.
+            use_upsample (bool):
+                Whether upsample at the end in self.forward.
+        """
+        super().__init__()
+
+        self.n_res_blocks = n_res_blocks
+        # define resblocks
+        self.resnet_layers = nn.ModuleList([])
+        self.attention_layers = nn.ModuleList([])
+        for res_block_i in range(n_res_blocks):
+            self.resnet_layers.append(
+                ResBlock(
+                    in_channels=(in_channels if res_block_i == 0 else out_channels),
+                    out_channels=out_channels,
+                )
+            )
+
+        # prepare upsample layer if needed
+        self.conv_for_upsample = nn.Conv2d(out_channels, out_channels, 3, padding=1) if use_upsample else None
+
+    def forward(
+        self,
+        x: torch.Tensor,
+    ) -> torch.Tensor:
+        """apply this Decoder block.
+
+        Args:
+            x (torch.Tensor):
+                Input features.
+                (shape: [batch_size, in_channels, height, width])
+        Returns:
+            torch.Tensor
+                Output features.
+                    If use_upsample is True, shape is
+                    [batch_size, out_channels, height*2, width*2],
+                    otherwise [batch_size, out_channels, height, width].
+        """
+        # apply resblocks
+        feature = x
+        for i in range(self.n_res_blocks):
+            feature = self.resnet_layers[i](feature)
+        # upsamples features if needed
+        if self.conv_for_upsample is not None:
+            feature = nn.functional.interpolate(feature, scale_factor=2, mode="nearest")
+            feature = self.conv_for_upsample(feature)
+        return feature
+
+
+class LatentVisualizationDecoder(nn.Module):
+    """Decoder for visualizing latents as image."""
+
+    def __init__(
+        self,
+        input_n_patches: size_2d,
+        input_latents_dim: int,
+        decoder_blocks_in_and_out_channels: list[tuple[int, int]],
+        n_res_blocks: int,
+        num_heads: int,
```

Comment:
機能要望。出力サイズを任意に指定できるようにしてほしい。
[`torchvision.transforms.functional.center_crop`](https://pytorch.org/vision/main/generated/torchvision.transforms.functional.center_crop.html)を用いて実装すると楽だと思う。
Convolutionによって出力される推定画像サイズが output sizeよりも下回っていた場合は Assertionエラーを出すと良いと思う。
```suggestion
        num_heads: int,
        output_size: size_2d
```

Reply:
@zassou65535 
> LatentVisualizationDecoderの__init__の引数として、「元の入力(n_patches_height, n_patches_width)を2の何乗倍したいか」を渡す

これはやってもやらなくてもいいかな。出力サイズが直感的に推定できるのは良いと思うから、やっていいとは思う。


> 「出力Tensorのサイズを任意に設定できます！！」からの「実は出力をcropしてるだけでした」って実装あんまり良くない気がする
> → ネットワークの構造に全く依存しない箇所で無理やり出力を決めてしまっている所があまり良くない気がした

別に問題ないと思う。`output_size`っていうのが、Convolutionを組み合わせて導かれるような意図を持ってしまうなら、確かに問題にはなり得るけど、Convolutionで指定よりも大きなサイズまでUpsamplingした上で、Cropするという意図が明確に伝わるコードであれば、全く問題にならないと思う。
あと出力をただCropするだけだからハイパラに影響があるとは考えられないな。

 `output_size`はデフォルトNoneにして置いて、設定されている時だけその処理を実行するっていうのが良いかもね。

Reply:
> 入力patch(batch_size, n_patches, dims)に対してtorch.meanを実行

これは避けたいな... I-JEPAは可視化に関してあまり詳細に追及していないので、信頼しきれない。

Reply:
ごめんちょっと質問がずれるんだけど、出力画像のサイズって任意で指定できなきゃダメ？`assert 2の累乗かどうか(output_size)`みたいに、例えば2の倍数のみ指定可能みたいな制約ってつけたらまずい？

Reply:
@zassou65535 このモジュール自体にその機能を実装する必要はないと思う。結局画像の再構成誤差を取る際に 144x144の形に直さないといけないと思うから、それが問題なく実装できればなんでも良いと思う。ただ、Trainerで任意の画像サイズにCropしたりする処理を書くのはお勧めできないな。出力サイズのプロパティは常にモデルが持つものなので、そこで設定できるようにするのが良い。`nn.Sequential`で、CenterCropをするレイヤーと結合しても良いかもね。

Reply:
どっちかっていうと必要不必要というよりは、この制約があった方がモデルの実装が単純化できそうという意図があって聞いてみた感じだった
あと、出力画像って絶対I-JEPAへの入力サイズと同一じゃなきゃだめ？I-JEPAへの入力は144x144だけど、Decoderからの出力は128x128ですとかはダメ？

Reply:
@zassou65535 その発想はなかった。ターゲット画像をモデルの出力画像のサイズに合わせてリサイズするってことね。

まあそれなら良いと思う。Approveするよ。

Reply:
OKじゃあその方針で実装しよう　Approveしてもらった後でアレなんだけど、これに関して一回実装を少し改良するからその時はもう1度レビューしてもらってもいい？

Reply:
@zassou65535 金曜あたりまでに TrainerのPR出せるならOK. 
もし単に実装を綺麗にしたいというエンジニアのエゴ的なものなら程々に。すでに要件は満たしているからね。